idaes/core/initialization/initializer_base.py

```diff
@@ -0,0 +1,399 @@
+        try:
+            results = self.initialization_routine(model)
+        # 6. finally: Restore model state
+        finally:
+            self.restore_model_state(model)
+
+        # 7. Check convergence
+        return self.postcheck(model, results_obj=results)
+
+    def get_current_state(self, model: Block):
+        """
+        Get and store current state of variables (fixed/unfixed) and constraints/objectives
+        activated/deactivated in model.
+
+        Args:
+            model: Pyomo model to get state from.
+
+        Returns:
+            dict serializing current model state.
+        """
+        self.initial_state = to_json(model, wts=StoreState, return_dict=True)
+
+        return self.initial_state
+
+    def load_initial_guesses(
+        self, model: Block, initial_guesses: dict = None, json_file: str = None
+    ):
+        """
+        Load initial guesses for variables into model.
+
+        Args:
+            model: Pyomo model to be initialized.
+            initial_guesses: dict of initial guesses to load.
+            json_file: file name of json file to load initial guesses from as str.
+
+        Note - can only provide one of initial_guesses or json_file.
+
+        Returns:
+            None
+
+        Raises:
+            ValueError if both initial_guesses and json_file are provided.
+        """
+        if initial_guesses is not None and json_file is not None:
+            self.status = InitializationStatus.Error
+            raise ValueError(
+                "Cannot provide both a set of initial guesses and a json file to load."
+            )
+
+        if initial_guesses is not None:
+            self._load_values_from_dict(model, initial_guesses)
+        elif json_file is not None:
+            # TODO: What should we load here?
+            # As this is just initialization, for now I am only loading variable values
+            from_json(
+                model, fname=json_file, wts=StoreSpec().value(only_not_fixed=True)
+            )
+        else:
+            self.get_logger(model).info_high(
+                "No initial guesses provided during initialization."
+            )
+
+    def fix_initialization_states(self, model: Block):
+        """
+        Call to model.fix_initialization_states method. Method will pass if
+        fix_initialization_states not found.
+
+        Args:
+            model: Pyomo Block to fix states on.
+
+        Returns:
+            None
+        """
+        try:
+            model.fix_initialization_states()
+        except InitializationError:
+            pass
+
+    def precheck(self, model: Block):
+        """
+        Check for satisfied degrees of freedom before running initialization.
+
+        Args:
+            model: Pyomo Block to fix states on.
+
+        Returns:
+            None
+
+        Raises:
+            InitializationError if Degrees of Freedom do not equal 0.
+        """
+        if not degrees_of_freedom(model) == 0:
```

Comment:
Not a big deal, but on large models the `degrees_of_freedom` call could take some noticeable amount of time.  You may want to store it in case you need it again in the message.

Reply:
I considered that, but didn't because I was hoping for many of the initializers to be general purpose and to be reused within a single workflow. E.g. create one instance of the Block Triangularization approach and then use it for multiple models.

Due to this, I wanted to avoid saving model specific data wherever possible (such as degrees of freedom). That said, there is enough model specific information that gets saved as it is (e.g. initialization status, initial state) that maybe it doesn't matter. We could also look at indexing all of these by `model.name` so that it is always clear what data applies to which model.